rust/lance/src/dataset/schema_evolution.rs

```diff
@@ -303,6 +315,79 @@ async fn add_columns_impl(
     Ok(fragments)
 }
 
+async fn add_columns_from_stream(
+    dataset: &Dataset,
+    mut stream: SendableRecordBatchStream,
+    schemas: Option<(Schema, Schema)>,
+    batch_size: Option<u32>,
+) -> Result<Vec<Fragment>> {
+    let fragments = dataset.get_fragments();
+    let mut new_fragments = Vec::with_capacity(fragments.len());
+    let mut last_seen_batch: Option<RecordBatch> = None;
+    for fragment in fragments {
+        let mut updater = fragment
+            .updater::<String>(Some(&[]), schemas.clone(), batch_size)
+            .await?;
+        while let Some(batch) = updater.next().await? {
+            debug_assert_eq!(batch.num_columns(), 1);
+            let mut rows_remaining = batch.num_rows();
+
+            let mut batches = Vec::new();
+            if let Some(last_batch) = last_seen_batch.take() {
+                if last_batch.num_rows() > rows_remaining {
+                    let new_batch = last_batch.slice(0, rows_remaining);
+                    batches.push(new_batch);
+                    last_seen_batch = Some(
+                        last_batch.slice(rows_remaining, last_batch.num_rows() - rows_remaining),
+                    );
+                    rows_remaining = 0;
+                } else {
+                    rows_remaining -= last_batch.num_rows();
+                    batches.push(last_batch);
+                    last_seen_batch = None;
+                }
+            }
+
+            while rows_remaining > 0 {
+                let next_batch = stream.next().await.ok_or_else(|| {
```

Comment:
The next batch can be last seen batch or new from the stream, then we can avoid these dup lines